readme
The ACM Algorithms Policy says, in part,

Dissemination Agreement

Submittal of an algorithm for publication in one of the ACM
Transactions implies that unrestricted use of the algorithm
within a computer is permissible.  General permission to
copy and distribute the algorithm without fee is granted
provided that the copies are not made or distributed for
direct commercial advantage.  The ACM copyright notice and
the title of the publication and its date appear, and
notice is given that copying is by permission of the
Association for Computing Machinery.  To copy otherwise,
or to republish, requires a fee and/or specific permission.


ACM sells the journal and the algorithms (separately).  For
current price and ordering information, look at an issue of
the journal in your library or contact
Association for Computing Machinery
P.O.Box 12105
Church Street Station
New York NY 10249
(212)869-7440


Changes
April 22 2021
   * Added 1013-1014 from Volume 47, number 1, (January 2021)
   * Added 1012 from Volume 46, number 4, (November 2020)
October 19 2020
   * Added 1011 from Volume 46, number 3, (September 2020)
August 7 2020
   * Added 1003-1006 from Volume 46, number 1, (April 2020)
   * Added 1007-1010 from Volume 46, number 2, (June 2020)
January 12 2020
   * Updated 999 at the request of the author -- updated user manual supplied.
January 7 2020
   * Added 999-1002 from Volume 45, number 4, (December 2019)
October 13 2019
   * Added 995-998 from Volume 45, number 3, (August 2019)
August 8 2019
   * Added 993 and 994 from Volume 45, number 2, (June 2019)
August 7 2019
   * Added 991 and 992 from Volume 45, number 1, (February 2019)
February 1 2019
   * Added 989 and 990 from Volume 44, number 3, (August 2018)
May 8 2018
   * Added 987 and 988 from Volume 44, number 3, (April 2018)
October 13 2017
   * Added 986 from Volume 44, number 2, (October 2017)
   * Corrected publication month for Algs 979-985 (now Oct).
September 30 2017
   * Added 979-985 from Volume 44, number 2, (September 2017)
   * Added 975-978 from Volume 44, number 1, (July 2017)
April 21 2017
   * Added 973, 974 and Remark on 936 from Volume 43, number 4 
     (March 2017)
February 3 2017
   * Added 968-972 from Volume 43, number 3 (January 2017)
September 16 2016
   * Added 965 from Volume 43, number 1 (August 2016)
   * Added 966, 967 from Volume 43, number 2 (September 2016)
   * Removed files 620 and 620.bib (all other algorithm files
     are only available as compressed files.
   * Replaced index.gz and index.html.gz by .zip files
September 14 2016:
   * Added 100.zip and 101.zip. These files are identical as
   the two algorithms are inter-related and the driver program
   uses both codes.
August 30 2016:
   * Added 963-964 from Volume 42, number 4 (July 2016)
July 23 2016
   * Updated contact information on CALGO web pages.
June 19 2016:
   * Added 960-962 from Volume 42, number 3 (June 2016)
   * Updated 916 to include the code submitted with Remark 
     in Volume 42, number 3 (Jun 2016)
June 15 2016:
   * Added 957 from Volume 42, number 1 (February 2016)
   * Added 958-959 from Volume 42, number 2 (June 2016)
May 30 2016
   * Added 955-956 from Volume 42, number 1 (February 2016)
May 27 2016:
   * Added new code supplied as Remark to Algorithm 673
     appearing in Volume 42, number 1.
December 22 2015:
   * Added 947-954 from Volume 41, numbers (1-4).
August 29 2015:
   * Mended a number of odd/broken compressed files that were on
     the CALGO site:
       755.gz, 756.gz, 823.gz replaced by zip files
       709.gz, 709a.gz, 709b.gz replaced by 709.zip
       573.gz, 573d.gz, 573s.gz replaced by 573.zip
       358.gz, 805.gz, 830.gz, 832.zip all updated
       620.zip added -- this contains all the 620 components.
September 26 2014
   * Updated 897 following accepted remark.
September 1 2014
   * Added 943-946 from Volume 40, number 4 (June 2014)
May 17 2014
   * Fixed problem with Algorithm 917 missing from indexes.
May 5 2014
   * Added 935-938 from Volume 40, number 2 (February 2014)
   * Added 939-942 from Volume 40, number 3 (April 2014)
October 9 2013
   * Added 929, 930 from volume 39, number 4 (July 2013)
   * Added 931-934 from volume 40, number 1 (October 2013)
July 18 2013
   * Added 928 from volume 39, number 3 (May 2013)
   * Updated 905 from code supplied by author thackerw@winthrop.edu
March 3 2013
   * Added 926-927 from volume 39, number 2 (February 2013)
January 25 2013
   * Replaced 923 after editing Licence file -- author request
January 3 2013
   * Added 924-925 from volume 39, number 1 (November 2012)
September 10 2012
   * Added 920-923 from volume 38, number 4 (August 2012)
June 17 2012
   * Added 917-919 from volume 38, number 3 (April 2012)
March 14 2012
   * Corrected typo in index files (Alg 916)
February 29 2012
   * Replaced 916 -- old version of code uploaded in error.
February  5 2012
   * Added 916 from Volume 38, number 2 (December 2012)
December 30 2011
   * Replaced 905 due to corrupt blas.f file in the original (pointed
     out by Ronald.M.Bloom@aero.org).
December 7 2011
   * Added 913-915 from Volume 38, number 1 (November 2011)
   * Updates 778 as per Remark in Volume 38, No. 1
April 12 2011
   * Added 909-911 from volume 37, number 4 (April 2011)
October 24 2011
   * Changed Algorithm 419 in line with changes from Remark on Algorithm 419 by
     David H. Withers CACM (March 1974) Vol 17 No 3 p. 157
     Pointed out by j.a.horsmeier@wanadoo.nl (Jos)
October 11 2010
   * Added 904-908 from volume 37, number 3 (September 2010)
May  20 2010
   * Added 902, 903 from volume 37, number 2 (April 2010)
Feb  27 2010
   * Added 898, 899, 900, 901 from volume 37, number 1 (January 2010)
July 29 2009
   * Added 895, 896, 897 from volume 36, number 3 (July 2009)
Apr  6 2009
   * Added 894 from volume 36, number 2 (March 2009)
Mar  30 2009
   * Added 891, 892, 893 from volume 36, number 1 (March 2009)
Feb  12 2009
   * Added 890 from volume 35, number 4 (February 2009)
Jan   6 2009
   * Added 885-889 from volume 35, number 3 (October 2008)
Sept  4 2008
   * Added 881-884 from volume 35, number 2 (July 2008)
July 25 2008
   * Added 876-877 from volume 34, number 4 (July 2008)
   * Added 878-880 from volume 35, number 1 (July 2008)
June 11 2008
   * Added 873-875 from volume 34, number 2 and 3 (March and May 2008)
Feb   8 2008
   * Added 867-869 from volume 33, numbers 3 and 4 (August 2007)
   * Added 870-872 from volume 34, number 1 (January 2008)
Aug   3 2007
   * Added 863-866 from volume 33, numbers 1 and 2.
July  3 2007
   * Updated 620 to include GAMS classifications to algorithms 814-862 inc.
Jan  10 2007
   * Updated 644 as a result of an accepted remark by Kodama
Jan   5 2007
   * Added 857-862 from volume 32, 4 (December 2006)
Nov  21 2006
   * Added 358 -- donated by John Burkardt <burkardt@csit.fsu.edu>.
Oct   5 2006
   * Added 854 from volume 32, 2 (June 2006)

796/main.f
C   DRIVER PROGRAM FOR TESTING SOFTWARE INVLTF  IMPLEMENTING           
C                                                                      
C   FOURIER BASED METHOD FOR THE NUMERICAL INVERSION                   
C                                                                      
C                  OF LAPLACE TRANSFORMS                              
C                                                                    
C                                                                 
C                                                              
C                                                           
C                     JANUARY , 1998      
C                                                                      
C                                                                      
C   AUTHORS: D'AMORE LUISA, GIULIANO LACCETTI, ALMERICO MURLI            
C                                                                      
C                                                                      
C
C
C
C   REFERENCES
C   ==========
C
C
C
C
C
C
C
C
C
C
C   REFERENCES
C   ==========
C
C   D'AMORE L., LACCETTI G., MURLI A.,  -    "ALGORITHM XXX: A FORTRAN
C      SOFTWARE PACKAGE FOR THE NUMERICAL INVERSION OF THE
C      LAPLACE TRANSFORM BASE ON FOURIER SERIES' METHOD"
C
      PROGRAM MAIN
C     ***********************************************************
C     DRIVER  PROGRAM TO TEST  THE ROUTINE INVLTF
C     FOR THE INVERSION OF A LAPLACE TRANSFORM FUNCTION.
C     THIS VERSION USES BOTH REAL AND COMPLEX DOUBLE PRECISION
C     OPERATIONS.
C     THE MAIN PROGRAM ALLOWS THE INVERSION OF A SET OF LAPLACE
C     TRANSFORM FUNCTIONS WHICH CAN BE ADDRESSED  BY A NATURAL
C     NUMBER  BETWEEN  1 AND 34.
C     FOR THE COMPLETE LIST SEE THE TABLE IN THE COMPANION PAPER.
C
C     THIS IS A SELF-CONTAINED DRIVER FOR THE INVLTF ROUTINE
C     COMPRISING A MAIN PROGRAM AND SUBPROGRAMS QDACC, BACKCF,
C     AND THE LAPLACE TRANSFORM PAIRS FZ AND FEX
C
C     THIS DRIVER ALLOWS TO OBTAIN UP TO 50 VALUES OF THE
C     INVERSE  LAPLACE FUNCTION 
C     
C     ************************************************************
C
C
C
C     .. Parameters ..
      INTEGER NMAX,FMAX,LASTFN
      PARAMETER (NMAX=550,FMAX=50,LASTFN=34)
C     ..
C     .. Scalars in Common ..
      INTEGER NFUN
C     ..
C     .. Local Scalars ..
      DOUBLE PRECISION EXF,SIGMA0,TOL,SSBAR
      INTEGER I,NT
      CHARACTER*100 AA,BB
      LOGICAL STOP
C     ..
C     .. Local Arrays ..
      DOUBLE COMPLEX WORK(2,0:2*NMAX)
      DOUBLE PRECISION DIFABS(FMAX),DIFREL(FMAX),FZINV(FMAX),
     +                 TARRAY(FMAX),ERROR(3,FMAX),ERR(3)
      INTEGER IFAIL(FMAX),IFZEVAL(FMAX)
C     ..
C     .. External Functions ..
      DOUBLE COMPLEX FZ
      DOUBLE PRECISION FEX
      EXTERNAL FZ,FEX
C     ..
C     .. External Subroutines ..
      EXTERNAL INVLTF
C     ..
C     .. Intrinsic Functions ..
      INTRINSIC ABS
C     ..
C     .. Common blocks ..
      COMMON /NF/NFUN
C     ..
C     *****************************************************************
C                   SET UP THE OUTPUT
C     *****************************************************************

C
C Loop through all the tests
C
      DO 50 NFUN = 1,LASTFN

C
C Skip the tests that require the use of NAG library routines
C
C Commented out lines need to be reintroduced in the routines
C FZ and FEX after labels 9, 36, 37, 38
C
          IF (NFUN.EQ.9 .OR. NFUN.EQ.36 .OR. NFUN.EQ.37 .OR.
     +        NFUN.EQ.38) GO TO 50
          AA =
     +'*****************************************************************
     +***********************'
          BB =
     +      '<><><><><><><><><><><><><><><><><><><><><><><><><><><><><>'
          WRITE (*,FMT=9000) AA,AA,AA,AA
          WRITE (*,FMT=9090)
C     READ (*,FMT=*) NFUN
          WRITE (*,FMT=9050) NFUN
          WRITE (*,FMT=9110)
C     READ (*,FMT=*) SIGMA0
	  SIGMA0 = 0.0
          WRITE (*,FMT=9120) SIGMA0
	  IF (NFUN .EQ. 18) SIGMA0 = 3.0
          IF (NFUN .EQ. 23) SIGMA0 = 0.25D0
          IF (NFUN .EQ. 29) SIGMA0 = 2.D0
          WRITE (*,FMT=9080) BB,BB,BB,BB
C     *****************************************************************
C          NOW, SET INPUT PARAMETERS FOR INVLTF
C     *****************************************************************
C
C      NT IS THE NUMBER OF T-VALUES USED HERE FOR THE TEST  FUNCTIONS.
C      THE MAXIMUM ALLOWED IS THE DIMENSION OF THE ARRAY TVAL
C     *****************************************************************
C
C     IN THIS TEST PROGRAM THE INVERSE FUNCTION IS REQUESTED IN THE
C     FOLLOWING VALUES OF T:
C     T=1,20, STEP=0.5 AND T=20,65 STEP=5
C     REMARK:
C     ======
C     FOR T SMALL, THAT IS FOR  T=1,20, STEP=0.5, ALL THE RESULTS ARE
C     QUITE ACCURATE, WHILE FOR T LARGE, THAT IS FOR T=20,65 STEP=5,
C     IN SOME CASES THE RESULTS COULDN'T BE ACCURATE. IN SUCH CASES 
C     THE AUTHORS SUGGEST TO USE AN ASYMPTOTIC INVERSION METHOD. 
C     *****************************************************************

      NT = 44
      DO 10 I = 1,39
           TARRAY(I) =1.D0 + (I-1)*0.5D0
   10 CONTINUE
      TARRAY(40) = 30.
      DO 20 I = 41,44
          TARRAY(I) = 30.D0 + 5.D0* (I-40)
   20 CONTINUE
      TOL = .1D-5
      SSBAR=0.d0
      WRITE (*,FMT=9010) AA,AA
      WRITE (*,FMT=9030) AA,AA
      WRITE (*,FMT=9020) TOL
C
C ************************************************************
C                    CALL OF THE ROUTINE INVLTF
C ************************************************************
C
C
      DO 2 I=1,NT
        CALL INVLTF(TOL,TARRAY(I),FZ,SIGMA0,SSBAR,NMAX,FZINV(I),ERR,
     +            IFZEVAL(I),WORK,IFAIL(I))
        ERROR(1,I)= ERR(1)
        ERROR(2,I)= ERR(2)
        ERROR(3,I)= ERR(3)
2     CONTINUE
C
C
C
      WRITE (*,FMT=9040) AA,AA
     
      DO 30 I = 1,NT
      STOP= .FALSE.
      IF (IFAIL(I).NE. 0 )THEN
         WRITE (*,FMT=9100) I,IFAIL(I)
         STOP=.TRUE.
      ENDIF
 

      IF (.NOT. STOP) THEN    

C
              EXF = FEX(TARRAY(I))
              DIFABS(I) = ABS(EXF-FZINV(I))
              IF (EXF.NE.0.D0) THEN
                  DIFREL(I) = DIFABS(I)/ABS(EXF)

              ELSE
                  DIFREL(I) = DIFABS(I)
              END IF

              EXF = FEX(TARRAY(I))
              IF (IFAIL(I).NE.-2) THEN
                  WRITE (*,FMT=9070) TARRAY(I),FZINV(I),EXF,ERROR(1,I),
     +         DIFREL(I),ERROR(3,I),ERROR(2,I),DIFABS(I),IFZEVAL(I),
     +         IFAIL(I)

              ELSE
                  WRITE (*,FMT=9060) TARRAY(I),IFAIL(I)
              END IF
      END IF

   30     CONTINUE
   50     CONTINUE

      STOP
C
 9000 FORMAT (1X,A45,A45,/,/,15X,
     +       '           SUBROUTINE INVLTF             ',/,15X,
     +       'NUMERICAL INVERSION OF A LAPLACE TRANSFORM:',/,15X,
     +       ' THIS VERSION USES BOTH REAL AND COMPLEX',/,15X,
     +       ' DOUBLE PRECISION OPERATIONS',/,/,1X,A45,A45,/)
 9010 FORMAT (/,A45,A45,/,/,16X,'        OUTPUT',/)
 9020 FORMAT (/,1X,'TOLL --> ',E15.7)
 9030 FORMAT (1X,'T      : POINT AT WHICH THE INVERSE TRANSFORM IS',
     +       ' COMPUTED;',/,/,1X,
     +       'FEX    : EXACT VALUE OF THE INVERSE TRANSFORM;',/,/,1X,
     +       'FCAL   : COMPUTED VALUE OF THE INVERSE TRANSFORM;',/,/,1X,
     +       'ESTREL : ESTIMATED RELATIVE ERROR;',/,/,1X,
     +       'RELERR : ACTUAL RELATIVE ERROR;',/,/,1X,
     +       'ESTABS : ESTIMATED ABSOLUTE ERROR ;',/,/,1X,
     +       'ABSERR : ACTUAL ABSOLUTE ERROR;',/,/,1X,
     +       'N      : # OF FUNCTION EVALUATIONS;',/,/,1X,
     +       'IFAIL  : = 0 NO INPUT ERRORS; SUCCESSFUL RUN',
     +       '    (ACCURACY REACHED AND IFZEVAL<NMAX),',/,10X,
     +       '= 1 TOL >= 1,',/,10X,
     +       '= 2 VALT LESS THAN ZERO,',/,10X,
     +       '= -1 ACCURACY NOT REACHED AND IFZEVAL > NMAX;,',/,10X,
     +       '= -2 THE CHOICE FOR SSBAR  MAY BE NOT OPTIMAL;',/,10X,
     +       '        IN SUCH A CASE THE USER MAY SLIGHTLY',/,10X,
     +       '        CHANGE THE DEFAULT VALUE;',/,/,A45,A45,/,/)
 9040 FORMAT (/,A86,/,'   T',9X,'FCAL',9X,'FEX',8X,' ESTREL',3X,
     +       'RELERR',2X,'TRUNERR',2X,'ESTABS',3X,'ABSERR',3X,'N',2X,
     +       'IFAIL',/,A100,/,/)
 9050 FORMAT (/,' TEST FUNCTION ----->  ',I2,/)
 9060 FORMAT (F5.1,90X,I2)
 9070 FORMAT (F5.1,1X,E14.8,1X,E14.8,1X,E8.3,1X,E8.3,1X,E8.3,1X,E8.3,1X,
     +       E8.3,1X,I3,3X,I2)
 9080 FORMAT (/,/,1X,A45,A45,/,/,2X,
     +     '   THE T-VALUES AT WHICH THE INVERSE IS REQUIRED ARE T=1,20'
     +       ,' STEP=0.5 AND T=20,100 STEP=10.',/,/,1X,A45,A45)
 9090 FORMAT (1X,'TEST  FUNCTION : ')
 9100 FORMAT (/,1X,'ERROR DETECTED , I = ', I3, ' IFAIL=',I3)
 9110 FORMAT (/,' ABSCISSA OF CONVERGENCE  ---> ',/)
 9120 FORMAT (/,' ABSCISSA OF CONVERGENCE  ---> ',F5.1,/)
      END
